[a2av] 2D all-to-all-v exchange plan

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #168047
* #167792
* #167676
* #167144
* #166896
* #166899
* #166898
* #166774
* #163815
* #163814
* #166772
* #165482
* __->__ #165101
* #164181
* #164164

If we want an intra-node A2A kernel to run concurrently with the inter-node A2A kernel, the intra-node kernel would need  some meta information ahead of time, for example, the offset of a token in a destination rank. 

The `make2dExchangePlan` kernel in this PR prepares such information, and would be run ahead of the intra-node A2A kernel.

Re "the offset of a token in a destination rank":
I am assuming that the inter-node A2A is in a pull style, while the intra-node A2A is in a push style, so that they two kernels can connect better (e.g. via a signal). For push style, we would need the dest offset for each token.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -1007,6 +1007,150 @@ void _all_to_all_get(
+  int my_pe = nvshmem_team_my_pe(team);
+  int npes = nvshmem_team_n_pes(team);
+  int nsplits = ne * npes;
+  CUDA_KERNEL_ASSERT(npes <= MAX_N_PEERS && " Number of peers must be smaller than MAX_N_PEERS\n");
```
I thought the assert accepts 1 field only. Can double check.

ngimel commented on torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu:
```diff
@@ -1007,6 +1007,150 @@ void _all_to_all_get(
+    // experts I own
+    int expert_id = ne * my_pe + local_expert;
+    // out_splits is (ne, npes) flattened
+    out_splits[local_expert * npes + peer] = in_splits_all[peer][expert_id];
```
seem out_splits doesn't need to be symmetric?

kwen2501 committed:
Update on "[a2av] 2D all-to-all-v exchange plan"


If we want an intra-node A2A kernel to run concurrently with the inter-node A2A kernel, the intra-node kernel would need  some meta information ahead of time, for example, the offset of a token in a destination rank. 

The `make2dExchangePlan` kernel in this PR prepares such information, and would be run ahead of the intra-node A2A kernel.

Re "the offset of a token in a destination rank":
I am assuming that the inter-node A2A is in a pull style, while the intra-node A2A is in a push style, so that they two kernels can connect better (e.g. via a signal). For push style, we would need the dest offset for each token.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]
